File: Sean的leetcode/python/Read N Characters Given Read4.py
###
# Terminate condition:
# 1) read n characters
# 2) no character to read
# Time Complexity: O(n)
# Space Complexity: O(n)
###
# The read4 API is already defined for you.
# @param buf, a list of characters
# @return an integer
# def read4(buf):

class Solution(object):
    def read(self, buf, n):
        """
        :type buf: Destination buffer (List[str])
        :type n: Maximum number of characters to read (int)
        :rtype: The number of characters read (int)
        """
        idx = 0
        buf4 = [None] * 4
        while idx < n:
            buf_num = read4(buf4)
            buf_idx = 0
            while buf_idx < buf_num and idx < n:
                buf[idx] = buf4[buf_idx]
                buf_idx += 1
                idx += 1
            if buf_num < 4:
                return idx
        return idx


            

File: MyLeetCode/Leetcode/Read N Characters Given Read4.cpp
// Forward declaration of the read4 API.
int read4(char *buf);

class Solution {
public:
    /**
     * @param buf Destination buffer not the place of file
     * @param n   Maximum number of characters to read
     * @return    The number of characters read
     */
    int read(char *buf, int n) {
        int count = 0;
        while (count < n)
        {
            int r = read4(buf);
            count += r;
            buf += r;
            if (r < 4)
                break;
        }
        int len = min(count, n);
        return len;
    }
};